packit/local_project.py

```diff
@@ -53,124 +65,175 @@ def __init__(
-                change = True
-
-            if self.git_project and not self.repo_name:
-                self.repo_name = self.git_project.repo
-                change = True
-
-            if self.git_project and not self.namespace:
-                self.namespace = self.git_project.namespace
-                change = True
-
-            if self.git_repo and not self.git_url:
-                # this is prone to errors
-                # also if we want url to upstream, we may want to ask for it explicitly
-                # since this can point to a fork
-                # .urls returns generator
-                self.git_url = list(self.git_repo.remote().urls)[0]
-                logger.debug(f"remote url of the repo is {self.git_url}")
-                change = True
-
-            if self.git_repo and not self.namespace:
-                self.namespace = get_namespace(self.git_url)
-                logger.debug(f"namespace from url is {self.namespace}")
-                if self.namespace:
-                    change = True
+            # we are trying to get new information while it is possible
+            # new iteration is done only if there was a change in the last iteration
+
+            change = (
+                self._parse_repo_name_full_name_and_namespace()
+                or self._parse_git_repo_from_working_dir(offline=offline)
+                or self._parse_git_project_from_repo_namespace_and_git_project(offline)
+                or self._parse_git_service_from_git_project(offline)
+                or self._parse_ref_from_git_repo()
+                or self._parse_working_dir_from_git_repo()
+                or self._parse_git_repo_from_git_url(offline)
+                or self._parse_git_url_from_git_project(offline)
+                or self._parse_repo_name_from_git_project()
+                or self._parse_namespace_from_git_project()
+                or self._parse_git_url_from_git_repo()
+                or self._parse_namespace_from_git_repo()
+            )
 
         if ref:
+            # If ref was specified, we will checkout that
 
             if ref not in self.git_repo.branches:
                 self.git_repo.create_head(self._ref)
 
             self.git_repo.branches[self._ref].checkout()
 
+    @property
+    def ref(self):
+        """
+        Name of the HEAD if the HEAD is not detached,
+        else commit hash.
+        """
+        if self.git_repo:
+            return self._get_ref_from_git_repo()
+        return None
+
+    def _is_url(self, path_or_url):
+        try:
+            res = requests.head(path_or_url)
+            if res.ok:
+                return True
+            else:
+                logger.warning("path_or_url is nor directory nor url")
+        except requests.exceptions.BaseHTTPError:
+            logger.warning("path_or_url is nor directory nor url")
+        return False
+
+    def _parse_namespace_from_git_repo(self):
+        if self.git_repo and not self.namespace:
```

Comment:
What about to add there use case when it is called? like LocalProject(git_url, git_service=<Something>).
Or maybe even what command use it like `propose-update` uses two of them.

Reply:
I don't understand. If you want to see how the class is being utilized, just open pyčarm and press `alt+f7` and that's your most up-to-date answer.

Reply:
`alt+f7` is doing some funny window resizing in my case. What action do you have in mind?